From 1e9cba9338054a875a2af685eb9179d0246a0ed6 Mon Sep 17 00:00:00 2001
From: Brian Ma <chma0@nuvoton.com>
Date: Fri, 29 May 2020 10:51:11 +0800
Subject: [PATCH 3/4] Avoid power state always ADC cannot trigger alarm

---
 src/ADCSensor.cpp | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/src/ADCSensor.cpp b/src/ADCSensor.cpp
index 81383b0..394c5a7 100644
--- a/src/ADCSensor.cpp
+++ b/src/ADCSensor.cpp
@@ -239,6 +239,13 @@ void ADCSensor::checkThresholds(void)
     {
         return;
     }
+    else if (readState == PowerState::always)
+    {
+        thresholds::checkThresholds(this);
+    }
+    else
+    {
+        thresholds::checkThresholdsPowerDelay(this, thresholdTimer);
+    }
 
-    thresholds::checkThresholdsPowerDelay(this, thresholdTimer);
 }
-- 
2.17.1

