From b7996db2ce6006fe53806020e6911fc5f8c0ddcd Mon Sep 17 00:00:00 2001
From: Tim Lee <timlee660101@gmail.com>
Date: Tue, 10 Dec 2019 14:32:28 +0800
Subject: [PATCH] fix get property failed when executing channel access command

---
 user_channel/channel_mgmt.cpp | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/user_channel/channel_mgmt.cpp b/user_channel/channel_mgmt.cpp
index b56da33..3fd7faa 100644
--- a/user_channel/channel_mgmt.cpp
+++ b/user_channel/channel_mgmt.cpp
@@ -1244,18 +1244,15 @@ int ChannelConfig::getDbusProperty(const std::string& service,
                                    const std::string& property,
                                    DbusVariant& value)
 {
-    try
-    {
-        auto method =
-            bus.new_method_call(service.c_str(), objPath.c_str(),
+    auto method =
+        bus.new_method_call(service.c_str(), objPath.c_str(),
                                 "org.freedesktop.DBus.Properties", "Get");
 
-        method.append(interface, property);
+    method.append(interface, property);
 
-        auto reply = bus.call(method);
-        reply.read(value);
-    }
-    catch (const sdbusplus::exception::SdBusError& e)
+    auto reply = bus.call(method);
+
+    if (reply.is_method_error())
     {
         log<level::DEBUG>("get-property failed",
                           entry("SERVICE=%s", service.c_str()),
@@ -1264,6 +1261,9 @@ int ChannelConfig::getDbusProperty(const std::string& service,
                           entry("PROP=%s", property.c_str()));
         return -EIO;
     }
+
+    reply.read(value);
+
     return 0;
 }
 
-- 
2.17.1

