From aebd5ac48dca2dfa2404989a64bfc91b68decc67 Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Mon, 23 Dec 2019 15:30:33 +0800
Subject: [PATCH] add scale and change the number from uint16 to int64

Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 NodeManagerProxy.hpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/NodeManagerProxy.hpp b/NodeManagerProxy.hpp
index 963cf40..1a93dbd 100644
--- a/NodeManagerProxy.hpp
+++ b/NodeManagerProxy.hpp
@@ -603,9 +603,11 @@ class getNmStatistics : public Request
 
         iface->register_property("MaxValue", static_cast<double>(maxValue));
         iface->register_property("MinValue", static_cast<double>(minValue));
-        iface->register_property("Value", static_cast<uint16_t>(0));
+        iface->register_property("Value", static_cast<int64_t>(0));
         iface->register_property(
             "Unit", std::string("xyz.openbmc_project.Sensor.Value.Unit.Watts"));
+        iface->register_property(
+            "Scale", static_cast<int64_t>(0));
 
         iface->initialize();
     }
@@ -644,7 +646,7 @@ class getNmStatistics : public Request
                 dataReceived.data());
 
         iface->set_property(
-            "Value", static_cast<uint16_t>(getNmStatistics->data.stats.cur));
+            "Value", static_cast<int64_t>((int64_t)getNmStatistics->data.stats.cur));
     }
 
     void prepareRequest(uint8_t &netFn, uint8_t &lun, uint8_t &cmd,
-- 
2.17.1

