From 6058615aa2523e865e975ef41d3345b3ce357f8c Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Wed, 16 Oct 2019 17:52:55 +0800
Subject: [PATCH] activation: add applyTime support for non ubifs layout

Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 activation.cpp | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/activation.cpp b/activation.cpp
index 0c18821..b4d9073 100644
--- a/activation.cpp
+++ b/activation.cpp
@@ -200,7 +200,17 @@ auto Activation::activation(Activations value) -> Activations
         // Create active association
         parent.createActiveAssociation(path);
 
-        log<level::INFO>("BMC image ready, need reboot to get activated.");
+        if (Activation::checkApplyTimeImmediate() == true)
+        {
+            log<level::INFO>("Image Active. ApplyTime is immediate, "
+                             "rebooting BMC.");
+            Activation::rebootBmc();
+        }
+        else
+        {
+		    log<level::INFO>("BMC image ready, need reboot to get activated.");
+        }
+
         return softwareServer::Activation::activation(
             softwareServer::Activation::Activations::Active);
 #endif
-- 
2.17.1

