diff --git a/images.hpp b/images.hpp
index 97b8f7f..527c564 100644
--- a/images.hpp
+++ b/images.hpp
@@ -11,8 +11,12 @@ namespace image
 {
 
 // BMC flash image file name list.
-const std::vector<std::string> bmcImages = {"image-kernel", "image-rofs",
-                                            "image-rwfs", "image-u-boot"};
+const std::vector<std::string> bmcImages = {"image-kernel","image-kernel.sig",
+                                            "image-rofs","image-rofs.sig",
+                                            "image-rwfs","image-rwfs.sig",
+                                            "image-u-boot","image-u-boot.sig",
+                                            "publickey","publickey.sig"
+                                            };
 
 } // namespace image
 } // namespace software
