diff --git a/CMakeLists.txt b/CMakeLists.txt
index e4f3b3d..1d7ac34 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -6,7 +6,7 @@ set (CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake ${CMAKE_MODULE_PATH})
 set (DBUS_OBJECT_NAME "xyz/openbmc_project/Smbios/MDR_V2")
 set (DBUS_INTF_NAME "xyz.openbmc_project.Smbios.MDR_V2")
 
-set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-rtti")
+#set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-rtti")
 
 include_directories (
     ${CMAKE_CURRENT_SOURCE_DIR}/include/phosphor-dbus-interfaces
