Index: dropbear-2019.78/svr-authpam.c
===================================================================
--- dropbear-2019.78.orig/svr-authpam.c
+++ dropbear-2019.78/svr-authpam.c
@@ -277,6 +277,13 @@ void svr_auth_pam(int valid_user) {
 		send_msg_userauth_failure(0, 1);
 		goto cleanup;
 	}
+ //ref:https://github.com/h0tw1r3/dropbear
+ if ((rc = pam_open_session(pamHandlep, 0)) != PAM_SUCCESS) {
+		dropbear_log(LOG_WARNING, "pam_open_session() failed, rc=%d, %s",
+			rc, pam_strerror(pamHandlep, rc));
+		send_msg_userauth_failure(0, 1);
+		goto cleanup;
+	}
 
 	/* successful authentication */
 	dropbear_log(LOG_NOTICE, "PAM password auth succeeded for '%s' from %s",
