From 0e61eb0861588d0bfb83fd5a8600ee95889b23c7 Mon Sep 17 00:00:00 2001
From: kfting <kfting@nuvoton.com>
Date: Wed, 19 Dec 2018 13:35:04 +0800
Subject: [PATCH] meta-evb-npcm750: Add user password update support for LDAP

---
 obmc/wsgi/apps/rest_dbus.py | 39 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/obmc/wsgi/apps/rest_dbus.py b/obmc/wsgi/apps/rest_dbus.py
index 5f3a147..d68f134 100755
--- a/obmc/wsgi/apps/rest_dbus.py
+++ b/obmc/wsgi/apps/rest_dbus.py
@@ -1146,6 +1146,43 @@ class HostConsoleHandler(RouteHandler):
         ping_sender = Greenlet.spawn(send_ws_ping, wsock, WEBSOCKET_TIMEOUT)
         gevent.joinall([wsock_reader, sock_reader, ping_sender])
 
+class NuvoPasswordHandler(MethodHandler):
+
+    rules = ['/setpassword']
+
+    def __init__(self, app, bus):
+        super(NuvoPasswordHandler, self).__init__(
+            app, bus)
+
+    @staticmethod
+    def chpasswd(username, newPassword):
+        try:
+            serv = 'passwd'
+            return pamela.change_password(username, newPassword, serv)
+
+        except KeyError:
+            return False
+
+    def find(self, **kw):
+        pass
+
+    def setup(self, **kw):
+        pass
+
+    def do_post(self, **kw):
+        if request.path == '/setpassword':
+            return self.do_changepassword(**kw)
+
+    def do_changepassword(self, **kw):
+        request.parameter_list = request.json.get('data')
+        if len(request.parameter_list) != 2:
+            abort(400, 'Wrong parameters')
+
+        if not self.chpasswd(*request.parameter_list):
+            abort(400, 'Wrong parameters')
+
+        return 'okay'
+
 class HostKvmHandler(RouteHandler):
     verbs = ['GET']
     # Naming the route kvm, because the numbering will help
@@ -1938,6 +1975,7 @@ class App(Bottle):
         self.image_upload_put_handler = ImagePutHandler(self, self.bus)
         self.download_dump_get_handler = DownloadDumpHandler(self, self.bus)
         self.certificate_put_handler = CertificatePutHandler(self, self.bus)
+        self.nuvo_pwd_handler = NuvoPasswordHandler(self, self.bus)
         if self.have_wsock:
             self.event_handler = EventHandler(self, self.bus)
             self.host_console_handler = HostConsoleHandler(self, self.bus)
@@ -1960,6 +1998,7 @@ class App(Bottle):
         self.image_upload_put_handler.install()
         self.download_dump_get_handler.install()
         self.certificate_put_handler.install()
+        self.nuvo_pwd_handler.install()
         if self.have_wsock:
             self.event_handler.install()
             self.host_console_handler.install()
-- 
2.7.4

