From 6bb74fc7981e3d07d6c4831a27f4f19cef0a62e9 Mon Sep 17 00:00:00 2001
From: kfting <kfting@nuvoton.com>
Date: Wed, 17 Jul 2019 11:01:47 +0800
Subject: [PATCH] meta-evb-npcm750: Add user password update support for LDAP

Signed-off-by: kfting <kfting@nuvoton.com>
---
 obmc/wsgi/apps/rest_dbus.py | 38 ++++++++++++++++++++++++++++++
 1 file changed, 38 insertions(+)

diff --git a/obmc/wsgi/apps/rest_dbus.py b/obmc/wsgi/apps/rest_dbus.py
index 3199e11..06103e5 100755
--- a/obmc/wsgi/apps/rest_dbus.py
+++ b/obmc/wsgi/apps/rest_dbus.py
@@ -1149,6 +1149,42 @@ class HostConsoleHandler(RouteHandler):
         ping_sender = Greenlet.spawn(send_ws_ping, wsock, WEBSOCKET_TIMEOUT)
         gevent.joinall([wsock_reader, sock_reader, ping_sender])
 
+class NuvoPasswordHandler(MethodHandler):
+
+    rules = ['/setpassword']
+
+    def __init__(self, app, bus):
+        super(NuvoPasswordHandler, self).__init__(app, bus)
+
+    @staticmethod
+    def chpasswd(username, newPassword):
+        try:
+            serv = 'passwd'
+            return pamela.change_password(username, newPassword, serv)
+
+        except KeyError:
+            return False
+
+    def find(self, **kw):
+        pass
+
+    def setup(self, **kw):
+        pass
+
+    def do_post(self, **kw):
+        if request.path == '/setpassword':
+            return self.do_changepassword(**kw)
+
+    def do_changepassword(self, **kw):
+        request.parameter_list = request.json.get('data')
+        if len(request.parameter_list) != 2:
+            abort(400, 'Wrong parameters')
+
+        if not self.chpasswd(*request.parameter_list):
+            abort(400, 'Wrong parameters')
+
+        return 'okay'
+
 class HostKvmHandler(RouteHandler):
     verbs = ['GET']
     # Naming the route kvm, because the numbering will help
@@ -1941,6 +1977,7 @@ class App(Bottle):
         self.image_upload_put_handler = ImagePutHandler(self, self.bus)
         self.download_dump_get_handler = DownloadDumpHandler(self, self.bus)
         self.certificate_put_handler = CertificatePutHandler(self, self.bus)
+        self.nuvo_pwd_handler = NuvoPasswordHandler(self, self.bus)
         if self.have_wsock:
             self.event_handler = EventHandler(self, self.bus)
             self.host_console_handler = HostConsoleHandler(self, self.bus)
@@ -1963,6 +2000,7 @@ class App(Bottle):
         self.image_upload_put_handler.install()
         self.download_dump_get_handler.install()
         self.certificate_put_handler.install()
+        self.nuvo_pwd_handler.install()
         if self.have_wsock:
             self.event_handler.install()
             self.host_console_handler.install()
-- 
2.17.1

