Index: 1.0-r1/ipmbbridged.cpp
===================================================================
--- 1.0-r1.orig/ipmbbridged.cpp
+++ 1.0-r1/ipmbbridged.cpp
@@ -616,7 +616,7 @@ int main(int argc, char *argv[])
     int r = -1;
     std::array<uint8_t, ipmbMaxFrameLength> buffer{};
 
-    const char *ipmbI2cSlave = "/sys/bus/i2c/devices/4-1010/slave-mqueue";
+    const char *ipmbI2cSlave = "/sys/bus/i2c/devices/4-1058/slave-mqueue";
     const char *ipmbI2cMaster = "/dev/i2c-4";
 
     sdbusplus::server::interface::interface ipmiInterface(
Index: 1.0-r1/ipmbbridged.hpp
===================================================================
--- 1.0-r1.orig/ipmbbridged.hpp
+++ 1.0-r1/ipmbbridged.hpp
@@ -27,7 +27,7 @@
  * @brief Ipmb addresses (8bit)
  */
 constexpr uint8_t ipmbBmcSlaveAddress = 0x20;
-constexpr uint8_t ipmbRqSlaveAddress = 0x58;
+constexpr uint8_t ipmbRqSlaveAddress = 0x10;
 
 /**
  * @brief Ipmb outstanding requests defines
