From edd9796994cfec6265f5879b5877722a293fab9d Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Wed, 30 Oct 2019 10:26:22 +0800
Subject: [PATCH] http: increase httpreqbody to 35MB

---
 http/http_connection.h | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/http/http_connection.h b/http/http_connection.h
index ef42f846..c1d58d1e 100644
--- a/http/http_connection.h
+++ b/http/http_connection.h
@@ -241,8 +241,8 @@ typename std::enable_if<(N > 0)>::type
 static std::atomic<int> connectionCount;
 #endif
 
-// request body limit size: 30M
-constexpr unsigned int httpReqBodyLimit = 1024 * 1024 * 30;
+// request body limit size: 35M
+constexpr unsigned int httpReqBodyLimit = 1024 * 1024 * 35;
 
 template <typename Adaptor, typename Handler, typename... Middlewares>
 class Connection
@@ -259,7 +259,7 @@ class Connection
         timerQueue(timerQueueIn)
     {
         parser.emplace(std::piecewise_construct, std::make_tuple());
-        // Temporarily changed to 30MB; Need to modify uploading/authentication
+        // Temporarily changed to 35MB; Need to modify uploading/authentication
         // mechanism
         parser->body_limit(httpReqBodyLimit);
         req.emplace(parser->get());
-- 
2.17.1

