From a6c6ddc30d7f2e5f046783fff7da60cf34f613c1 Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Fri, 15 Nov 2019 17:09:56 +0800
Subject: [PATCH] bmcweb: chassis: add indicatorLED support

Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 redfish-core/lib/chassis.hpp | 110 +++++++++++++++++++++++++++++++++--
 1 file changed, 104 insertions(+), 6 deletions(-)

diff --git a/redfish-core/lib/chassis.hpp b/redfish-core/lib/chassis.hpp
index 2b098a93..1ae59e02 100644
--- a/redfish-core/lib/chassis.hpp
+++ b/redfish-core/lib/chassis.hpp
@@ -117,6 +117,74 @@ void getIntrusionByService(std::shared_ptr<AsyncResp> aResp,
         "xyz.openbmc_project.Chassis.Intrusion", "Status");
 }
 
+
+/**
+ * Retrieves physical Led properties over dbus
+ */
+void getPhysicalLedData(std::shared_ptr<AsyncResp> aResp, std::string path)
+{
+    std::size_t found = path.rfind('/');
+    std::string led_name;
+
+    if (found != std::string::npos)
+    {
+        led_name = path.substr(found + 1);
+    }
+    else
+    {
+        return;
+    }
+
+    std::string obj = "xyz.openbmc_project.LED.Controller." + led_name;
+
+    crow::connections::systemBus->async_method_call(
+        [aResp](const boost::system::error_code ec,
+                                        const PropertiesType &properties) {
+            if (ec)
+            {
+                BMCWEB_LOG_DEBUG << "DBUS response error " << ec;
+                messages::internalError(aResp->res);
+                return;
+            }
+            BMCWEB_LOG_DEBUG << "Got " << properties.size()
+                             << " led properties.";
+            std::string output;
+            for (const auto &property : properties)
+            {
+                if (property.first == "State")
+                {
+                    const std::string *s =
+                        std::get_if<std::string>(&property.second);
+                    if (nullptr != s)
+                    {
+                        BMCWEB_LOG_DEBUG << "Identify Led State: " << *s;
+                        const auto pos = s->rfind('.');
+                        if (pos != std::string::npos)
+                        {
+                            auto led = s->substr(pos + 1);
+                            for (const std::pair<const char *, const char *>
+                                     &p :
+                                 std::array<
+                                     std::pair<const char *, const char *>, 3>{
+                                     {{"On", "Lit"},
+                                      {"Blink", "Blinking"},
+                                      {"Off", "Off"}}})
+                            {
+                                if (led == p.first)
+                                {
+                                    aResp->res.jsonValue["IndicatorLED"] = p.second;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        },
+        obj, path,
+        "org.freedesktop.DBus.Properties", "GetAll",
+        "xyz.openbmc_project.Led.Physical");
+}
+
 /**
  * Retrieves physical security properties over dbus
  */
@@ -289,9 +357,41 @@ class Chassis : public Node
 
                     auto health = std::make_shared<HealthPopulate>(asyncResp);
 
+
+                    // Get the list of all sensors for this Chassis element
+                    std::vector<std::string> sensorPaths = {
+                        "/all_sensors",
+                        "/sensors"
+                    };
+
+                    for (const std::string sensorPath : sensorPaths)
+                    {
+                        crow::connections::systemBus->async_method_call(
+                            [health](const boost::system::error_code ec,
+                                    std::variant<std::vector<std::string>> &resp) {
+                                if (ec)
+                                {
+                                    return; // no sensors = no failures
+                                }
+                                std::vector<std::string> *data =
+                                    std::get_if<std::vector<std::string>>(&resp);
+                                if (data == nullptr)
+                                {
+                                    return;
+                                }
+                                health->inventory = std::move(*data);
+                            },
+                            "xyz.openbmc_project.ObjectMapper",
+                            path + sensorPath,
+                            "org.freedesktop.DBus.Properties", "Get",
+                            "xyz.openbmc_project.Association", "endpoints");
+                    }
+
+                    health->populate();
+
                     crow::connections::systemBus->async_method_call(
-                        [health](const boost::system::error_code ec,
-                                 std::variant<std::vector<std::string>>& resp) {
+                        [asyncResp](const boost::system::error_code ec,
+                                std::variant<std::vector<std::string>> &resp) {
                             if (ec)
                             {
                                 return; // no sensors = no failures
@@ -302,15 +402,13 @@ class Chassis : public Node
                             {
                                 return;
                             }
-                            health->inventory = std::move(*data);
+                            getPhysicalLedData(asyncResp, std::move((*data)[0]));
                         },
                         "xyz.openbmc_project.ObjectMapper",
-                        path + "/all_sensors",
+                        path + "/leds",
                         "org.freedesktop.DBus.Properties", "Get",
                         "xyz.openbmc_project.Association", "endpoints");
 
-                    health->populate();
-
                     if (connectionNames.size() < 1)
                     {
                         BMCWEB_LOG_ERROR << "Got 0 Connection names";
-- 
2.17.1

