From 82891632642541e93979bb306a1c88889c352348 Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Wed, 28 Oct 2020 17:45:03 +0800
Subject: [PATCH] fix bmcweb crash

---
 http/timer_queue.hpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/http/timer_queue.hpp b/http/timer_queue.hpp
index 5baf7bee..924d3ae2 100644
--- a/http/timer_queue.hpp
+++ b/http/timer_queue.hpp
@@ -31,6 +31,7 @@ class TimerQueue
         if (index < dq.size())
         {
             dq[index].second = nullptr;
+            dq.pop_back();
         }
     }
 
-- 
2.17.1

