From 93231767e76b3107963ab29954e9bc491289b883 Mon Sep 17 00:00:00 2001
From: Stanley Chu <yschu@nuvoton.com>
Date: Tue, 26 Jan 2021 14:52:03 +0800
Subject: [PATCH] create a dbus log entry when watchdog timeout occurs.

Signed-off-by: Stanley Chu <yschu@nuvoton.com>
---
 watchdog.cpp | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/watchdog.cpp b/watchdog.cpp
index 3b5356f..69145dd 100644
--- a/watchdog.cpp
+++ b/watchdog.cpp
@@ -9,6 +9,8 @@
 #include <sdbusplus/exception.hpp>
 #include <xyz/openbmc_project/Common/error.hpp>
 #include <xyz/openbmc_project/State/Host/server.hpp>
+#include <xyz/openbmc_project/Control/Boot/error.hpp>
+#include "phosphor-logging/elog-errors.hpp"
 
 namespace phosphor
 {
@@ -276,6 +278,12 @@ void Watchdog::timeOutHandler()
     }
     else
     {
+        using namespace xyz::openbmc_project::Control::Boot;
+        using error =
+            sdbusplus::xyz::openbmc_project::Control::Boot::Error::WatchdogTimedOut;
+        report<error>(WatchdogTimedOut::WATCHDOG_ACTION(convertForMessage(action).c_str()),
+                      WatchdogTimedOut::WATCHDOG_TIMER_USE(convertForMessage(expiredTimerUse()).c_str()),
+                      WatchdogTimedOut::WATCHDOG_INTERVAL(WatchdogInherits::interval()));
         log<level::INFO>(
             "watchdog: Timed out",
             entry("ACTION=%s", convertForMessage(action).c_str()),
-- 
2.17.1

