diff --git a/arch/arm/boot/dts/nuvoton-npcm750-buv-runbmc.dts b/arch/arm/boot/dts/nuvoton-npcm750-buv-runbmc.dts
index 604d9c41d169..0eae04c4b43a 100644
--- a/arch/arm/boot/dts/nuvoton-npcm750-buv-runbmc.dts
+++ b/arch/arm/boot/dts/nuvoton-npcm750-buv-runbmc.dts
@@ -182,11 +182,11 @@
 					};
 					rofs@800000 {
 						label = "rofs";
-						reg = <0x800000 0x1500000>;
+						reg = <0x800000 0x3500000>;
 					};
 					rwfs@1d00000 {
 						label = "rwfs";
-						reg = <0x1d00000 0x300000>;
+						reg = <0x3d00000 0x300000>;
 					};
 				};
 			};
