diff --git a/kcsbridged.cpp b/kcsbridged.cpp
index a37cf5f..3014bfb 100644
--- a/kcsbridged.cpp
+++ b/kcsbridged.cpp
@@ -112,8 +112,15 @@ class SmsChannel

         static constexpr const char pathBase[] =
             "/xyz/openbmc_project/Ipmi/Channel/";
+        std::string objpath;
+        auto pos = channel.find("ipmi-");
+        if (pos != std::string::npos) {
+            objpath = pathBase + channel.substr(pos+strlen("ipmi-"));
+        } else {
+            objpath = pathBase + channel;
+        }
         std::shared_ptr<sdbusplus::asio::dbus_interface> iface =
-            server->add_interface(pathBase + channel,
+            server->add_interface(objpath,
                                   "xyz.openbmc_project.Ipmi.Channel.SMS");
         iface->register_method("setAttention",
                                [this]() { return setAttention(); });
@@ -121,6 +128,7 @@ class SmsChannel
                                [this]() { return clearAttention(); });
         iface->register_method("forceAbort", [this]() { return forceAbort(); });
         iface->initialize();
+        clearAttention();
     }

     bool initOK() const
