From b0a20757e05c3a229fc2d61b93fbe54048cad388 Mon Sep 17 00:00:00 2001
From: Stanley Chu <yschu@nuvoton.com>
Date: Tue, 14 Jul 2020 14:02:43 +0800
Subject: [PATCH] using async method call to commit log

Signed-off-by: Stanley Chu <yschu@nuvoton.com>
---
 elog.cpp | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/elog.cpp b/elog.cpp
index 5cefe81..7567b50 100644
--- a/elog.cpp
+++ b/elog.cpp
@@ -1,5 +1,6 @@
 #include "config.h"
 
+#include <sdbusplus/asio/connection.hpp>
 #include <phosphor-logging/elog.hpp>
 #include <stdexcept>
 
@@ -52,8 +53,12 @@ void commit(const char* name)
     auto msg = _prepareMsg("Commit");
     uint64_t id = sdbusplus::server::transaction::get_id();
     msg.append(id, name);
-    auto bus = sdbusplus::bus::new_default();
-    bus.call_noreply(msg);
+    //auto bus = sdbusplus::bus::new_default();
+    //bus.call_noreply(msg);
+    using message = sdbusplus::message::message;
+    boost::asio::io_context io;
+    auto conn = std::make_shared<sdbusplus::asio::connection>(io);
+    conn->async_send(msg, [](boost::system::error_code ec, message& ret){});
 }
 
 void commit(const char* name, Entry::Level level)
-- 
2.17.1

