From 73af8bb5632d44a77f6a4f902bdd21f99fedd0f3 Mon Sep 17 00:00:00 2001
From: Stanley Chu <yschu@nuvoton.com>
Date: Fri, 17 Jul 2020 17:21:59 +0800
Subject: [PATCH] create thread to send dbus message to commit log

Signed-off-by: Stanley Chu <yschu@nuvoton.com>
---
 elog.cpp | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/elog.cpp b/elog.cpp
index 5cefe81..d9a4c36 100644
--- a/elog.cpp
+++ b/elog.cpp
@@ -1,5 +1,6 @@
 #include "config.h"
 
+#include <thread>
 #include <phosphor-logging/elog.hpp>
 #include <stdexcept>
 
@@ -46,14 +47,22 @@ auto _prepareMsg(const char* funcName)
         b.new_method_call(host.c_str(), OBJ_INTERNAL, IFACE_INTERNAL, funcName);
     return m;
 }
+void send_dbus_msg(sdbusplus::message::message msg)
+{
+    auto bus = sdbusplus::bus::new_default();
+    bus.call_noreply(msg);
+}
 
 void commit(const char* name)
 {
     auto msg = _prepareMsg("Commit");
     uint64_t id = sdbusplus::server::transaction::get_id();
     msg.append(id, name);
-    auto bus = sdbusplus::bus::new_default();
-    bus.call_noreply(msg);
+    //auto bus = sdbusplus::bus::new_default();
+    //bus.call_noreply(msg);
+
+    std::thread t(send_dbus_msg, msg);
+    t.detach();
 }
 
 void commit(const char* name, Entry::Level level)
-- 
2.17.1

