From 9141db5a4f14dc1b8b653c1da840ab70b3d251dd Mon Sep 17 00:00:00 2001
From: kfting <kfting@nuvoton.com>
Date: Thu, 28 May 2020 10:21:47 +0800
Subject: [PATCH] Add the gateway setting only when the non-dhcp address is
 available.

Signed-off-by: kfting <kfting@nuvoton.com>
---
 ethernet_interface.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ethernet_interface.cpp b/ethernet_interface.cpp
index 9b9348a..f146e79 100644
--- a/ethernet_interface.cpp
+++ b/ethernet_interface.cpp
@@ -880,7 +880,7 @@ void EthernetInterface::writeConfigurationFile()
             }
         }
 
-        if (manager.getSystemConf())
+        if (!addrs.empty() && manager.getSystemConf())
         {
             const auto& gateway = manager.getSystemConf()->defaultGateway();
             if (!gateway.empty())
-- 
2.17.1

